src/manifolds/Symplectic.jl

```diff
@@ -0,0 +1,889 @@
+                                        ∈ T_p\operatorname{Sp}(2n).
+````
+
+[^Bendokat2021]:
+    > Bendokat, Thomas and Zimmermann, Ralf:
+	> The real symplectic Stiefel and Grassmann manifolds: metrics, geodesics and applications
+	> arXiv preprint arXiv:2108.12447, 2021 (https://arxiv.org/abs/2108.12447)
+"""
+function inverse_retract!(M::Symplectic, X, p, q, ::CayleyInverseRetraction)
+    U_inv = lu(add_scaled_I!(symplectic_inverse_times(M, p, q), 1))
+    V_inv = lu(add_scaled_I!(symplectic_inverse_times(M, q, p), 1))
+
+    X .= 2 .* ((p / V_inv .- p / U_inv) .+ ((p + q) / U_inv) .- p)
+    return X
+end
+
+@doc raw"""
+    manifold_dimension(::Symplectic{n})
+
+Returns the dimension of the symplectic manifold
+embedded in ``ℝ^{2n \times 2n}``, i.e.
+````math
+    \operatorname{dim}(\operatorname{Sp}(2n)) = (2n + 1)n.
+````
+"""
+manifold_dimension(::Symplectic{n}) where {n} = (2n + 1) * n
+
+@doc raw"""
+    project_riemannian!(M::Symplectic{n, ℝ}, Y, p, X) where {n}
+
+Compute the projection of ``X ∈ R^{2n × 2n}`` onto ``T_p\operatorname{Sp}(2n, ℝ)`` w.r.t.
+the Riemannian metric ``g`` [`RealSymplecticMetric`](@ref).
+
+The closed form projection mapping is given by [^Gao2021riemannian]
+````math
+    \operatorname{P}^{T_p\operatorname{Sp}(2n)}_{g_p}(X) = pQ\operatorname{sym}(p^TQ^TX),
+````
+where ``\operatorname{sym}(A) = \frac{1}{2}(A + A^T)``.
+"""
+function project_riemannian!(::Symplectic, Y, p, X)
+    Q = SymplecticMatrix(p, X)
+
+    pT_QT_X = p' * Q' * X
+    symmetrized_pT_QT_X = (1 / 2) .* (pT_QT_X + pT_QT_X')
+
+    Y[:, :] = p * Q * (symmetrized_pT_QT_X)
+    return Y
+end
+
+@doc raw"""
+    project_riemannian_normal!(M::Symplectic{n, ℝ}, Y, p, X)
+
+Project onto the normal of the tangent space ``(T_p\operatorname{Sp}(2n))^{\perp_g}`` at
+a point ``p ∈ \operatorname{Sp}(2n)``,
+relative to the riemannian metric ``g`` [`RealSymplecticMetric`](@ref).
+That is,
+````math
+(T_p\operatorname{Sp}(2n))^{\perp_g} = \{Y \in \mathbb{R}^{2n \times 2n} :
+                        g_p(Y, X) = 0 \;\forall\; X \in T_p\operatorname{Sp}(2n)\}.
+````
+The closed form projection operator onto the normal space is given by [^Gao2021riemannian]
+````math
+\operatorname{P}^{(T_p\operatorname{Sp}(2n))\perp}_{g_p}(X) = pQ\operatorname{skew}(p^TQ^TX),
+````
+where ``\operatorname{skew}(A) = \frac{1}{2}(A - A^T)``.
+
+[^Gao2021riemannian]:
+    > Gao, Bin and Son, Nguyen Thanh and Absil, P-A and Stykel, Tatjana:
+    > Riemannian optimization on the symplectic Stiefel manifold,
+    > SIAM Journal on Optimization 31(2), pp. 1546-1575, 2021.
+    > doi [10.1137/20M1348522](https://doi.org/10.1137/20M1348522)
+"""
+function project_riemannian_normal!(::Symplectic, Y, p, X)
+    Q = SymplecticMatrix(p, X)
+
+    pT_QT_X = p' * Q' * X
+    skew_pT_QT_X = (1 / 2) .* (pT_QT_X .- pT_QT_X')
+
+    Y[:, :] = p * Q * skew_pT_QT_X
+    return Y
+end
+
+function Base.rand(M::Symplectic{n}, hamiltonian_norm::Number=1.0) where {n}
```

Comment:
```suggestion
function Base.rand(M::Symplectic{n}; hamiltonian_norm::Number=1.0) where {n}
```
Extending `Base.rand` correctly is quite tricky. `rand(some_manifold)` is not an API we currently have but it should be fine, although any further options should be introduced as keyword arguments.

Reply:
Ok good to know. I've mostly used it as an easy way of generating initial points for optimization routines, but it may be cluttering the namespace more than it's worth. Is it worth keeping these functions for the convenience?